core/sail/fts/lucene-spin/src/main/java/org/eclipse/rdf4j/lucene/spin/LuceneSpinSailConnection.java

```diff
@@ -0,0 +1,261 @@
+import org.eclipse.rdf4j.sail.lucene.LuceneSailBuffer.Operation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.eclipse.rdf4j.sail.lucene.LuceneSailBuffer;
+import org.eclipse.rdf4j.sail.lucene.LuceneSailConnection;
+import org.eclipse.rdf4j.sail.lucene.SearchIndex;
+
+/**
+ * This connection inherits Lucene index supporting methods from {@link LuceneSailConnection}.
+ * 
+ * @author sauermann
+ * @author christian.huetter
+ * @author github.com/jgrzebyta
+ */
+public class LuceneSpinSailConnection extends NotifyingSailConnectionWrapper {
+
+	private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+	private final SearchIndex luceneIndex;
+
+	/**
+	 * the buffer that collects operations
+	 */
+	final private LuceneSailBuffer buffer = new LuceneSailBuffer();
+
+	/**
+	 * The listener that listens to the underlying connection. It is disabled during clearContext operations.
+	 */
+	protected final SailConnectionListener connectionListener = new SailConnectionListener() {
+
+		@Override
+		public void statementAdded(Statement statement) {
+			// we only consider statements that contain literals
+			if (statement.getObject() instanceof Literal) {
+				if (statement == null)
+					return;
+				// we further only index statements where the Literal's datatype is
+				// accepted
+				Literal literal = (Literal)statement.getObject();
+				if (luceneIndex.accept(literal))
+					buffer.add(statement);
+			}
+		}
+
+		@Override
+		public void statementRemoved(Statement statement) {
+			// we only consider statements that contain literals
+			if (statement.getObject() instanceof Literal) {
+				if (statement == null)
+					return;
+				// we further only indexed statements where the Literal's datatype
+				// is accepted
+				Literal literal = (Literal)statement.getObject();
+				if (luceneIndex.accept(literal))
+					buffer.remove(statement);
+			}
+		}
+	};
+
+	/**
+	 * To remember if the iterator was already closed and only free resources once
+	 */
+	private final AtomicBoolean closed = new AtomicBoolean(false);
+
+	public LuceneSpinSailConnection(NotifyingSailConnection wrappedConnection, SearchIndex luceneIndex) {
+		super(wrappedConnection);
+		this.luceneIndex = luceneIndex;
+
+		/*
+		 * Using SailConnectionListener, see <a href="#whySailConnectionListener">above</a>
+		 */
+
+		wrappedConnection.addConnectionListener(connectionListener);
+	}
+
+	@Override
+	public synchronized void addStatement(Resource arg0, IRI arg1, Value arg2, Resource... arg3)
+		throws SailException
+	{
+		super.addStatement(arg0, arg1, arg2, arg3);
+	}
+
+	@Override
+	public void close()
+		throws SailException
+	{
+		if (closed.compareAndSet(false, true)) {
+			try {
+				super.close();
+			}
+			finally {
+				try {
+					luceneIndex.endReading();
+				}
+				catch (IOException e) {
+					logger.warn("could not close IndexReader or IndexSearcher " + e, e);
+				}
+				// remember if you were closed before, some sloppy programmers
```

Comment:
This comment is irrelevant per the AtomicBoolean check above, and it isn't sloppy programming, there are background resource leak fix mechanisms that legitimately call close without knowing if it has been called before.